File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.js
/*
 * TC: O(N)
 * SC: O(N)
*/


function solution(s){
    const stack = []

    for (const char of s) {
        if (char === '(') {
            stack.push(char)
        } else {
            // 닫는 괄호가 나왔을 때 stack이 비어 있으면 false
            if (stack.length === 0) return false
            stack.pop()
        }
    }
    return stack.length === 0
}

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.py
def solution(s):
    stack = []
    for bracket in s:
        if(len(stack) == 0):
            stack.append(bracket)
            continue
        if bracket == ")" and stack[-1] == "(":
            stack.pop()
        else:
            stack.append(bracket)

    return True if len(stack) == 0 else False

